pkg/agent/flowexporter/exporter/exporter.go

```diff
@@ -148,56 +148,166 @@ func NewFlowExporter(connStore *connections.ConntrackConnectionStore, records *f
-		exp.process.CloseConnToCollector()
-		exp.process = nil
-		return
+}
+
+// GetExpiryFromExpirePriorityQueue returns the earliest timestamp (active expiry
+// or idle expiry) from expire priority queue.
+func (exp *flowExporter) GetExpiryFromExpirePriorityQueue() time.Duration {
+	currTime := time.Now()
+	if exp.expirePriorityQueue.Len() > 0 {
+		// Get the minExpireTime of the top item in ExpirePriorityQueue.
+		expiryDuration := MinExpiryTime + exp.expirePriorityQueue.MinExpireTime(0).Sub(currTime)
+		if expiryDuration < 0 {
+			return MinExpiryTime
+		}
+		return expiryDuration
+	}
+	if exp.activeFlowTimeout < exp.idleFlowTimeout {
+		return exp.activeFlowTimeout
 	}
-	klog.V(2).Infof("Successfully exported IPFIX flow records")
+	return exp.idleFlowTimeout
+}
+
+func (exp *flowExporter) sendFlowRecords() error {
+	exp.expirePriorityQueue.Mutex.Lock()
+	defer exp.expirePriorityQueue.Mutex.Unlock()
+
+	if exp.expirePriorityQueue.Len() == 0 {
+		return nil
+	}
+	currTime := time.Now()
+	for exp.expirePriorityQueue.Len() > 0 {
+		topItem := exp.expirePriorityQueue.Peek()
+		// If the top item is not timeout, then we do not have to check the following items.
+		if topItem.ActiveExpireTime.After(currTime) && topItem.IdleExpireTime.After(currTime) {
+			break
+		}
+		pqItem := heap.Pop(exp.expirePriorityQueue).(*flowexporter.ItemToExpire)
+		// The pqItem contains a flow record.
+		if !pqItem.IsDenyConn {
+			if pqItem.IsActive {
+				// Send a flow record if it is active and time out.
+				// TODO: more records per data set will be supported when go-ipfix supports size check when adding records
+				if err := exp.addConnToSet(pqItem.Conn, false); err != nil {
+					return err
+				}
+				if _, err := exp.sendDataSet(); err != nil {
+					return err
+				}
+				exp.numDataSetsSent = exp.numDataSetsSent + 1
+			}
+			if flowexporter.IsConnectionDying(&pqItem.Conn) {
+				// If the connection is in dying state or connection is not in conntrack
+				// table, we set the DyingAndDoneExport flag to do the deletion later.
+				pqItem.DyingAndDoneExport = true
+			} else {
+				if pqItem.IdleExpireTime.Before(currTime) && (pqItem.Conn.OriginalPackets <= pqItem.PrevPackets) && (pqItem.Conn.ReversePackets <= pqItem.PrevReversePackets) {
+					pqItem.IsActive = false
+				}
+				exp.ValidateAndUpdateStats(pqItem)
+				if pqItem.IdleExpireTime.Before(currTime) {
+					pqItem.IdleExpireTime = currTime.Add(exp.idleFlowTimeout)
+				}
+				if pqItem.ActiveExpireTime.Before(currTime) {
+					pqItem.ActiveExpireTime = currTime.Add(exp.activeFlowTimeout)
+				}
+				heap.Push(exp.expirePriorityQueue, pqItem)
+			}
+			klog.V(4).InfoS("Record for connection sent successfully", "flowKey", pqItem.ConnKey, "connection", pqItem.Conn)
+		}
+		// The pqItem contains a deny connection.
+		if pqItem.IsDenyConn {
+			if err := exp.addConnToSet(pqItem.Conn, true); err != nil {
```

Comment:
we should skip sending the record if `DyingAndDoneExport` right? 
I see the deletion period to be one minute. If that is the case, we do not want to send the repeated record.
We can do this check for conntrack connection too.

Reply:
The record will not be push back to the PQ if `DyingAndDoneExport` . Same for conntrack connection.